pkg/controller/core.oam.dev/v1alpha2/application/apply.go

```diff
@@ -93,108 +69,49 @@ func (h *appHandler) handleErr(err error) (ctrl.Result, error) {
+	appRev := h.currentAppRev
+	if h.app.Spec.Workflow != nil || appRev.Annotations[oam.AnnotationAppRevisionOnly] == "true" {
+		return h.createResourcesConfigMap(ctx, appRev, comps, policies)
+	}
+	if appWillRollout(h.app) {
+		return nil
 	}
 
-	owners := []metav1.OwnerReference{{
-		APIVersion: v1beta1.SchemeGroupVersion.String(),
-		Kind:       v1beta1.ApplicationKind,
-		Name:       h.app.Name,
-		UID:        h.app.UID,
-		Controller: pointer.BoolPtr(true),
-	}}
-	ac.SetOwnerReferences(owners)
-	var err error
+	var latestTracker *v1beta1.ResourceTracker
+	if h.app.Status.LatestRevision != nil {
+		latestTracker = &v1beta1.ResourceTracker{}
+		latestTracker.SetName(dispatch.ConstructResourceTrackerName(h.app.Status.LatestRevision.Name, h.app.Namespace))
+	}
+	// only do GC when ALL resources are dispatched successfully
+	// so skip GC while dispatching addon resources
+	d := dispatch.NewAppManifestsDispatcher(h.r.Client, appRev).EnableUpgradeAndSkipGC(latestTracker)
+	// dispatch workload addon resources before dispatching assembled manifests
 	for _, comp := range comps {
-		comp.SetOwnerReferences(owners)
-
-		if h.isNewRevision && h.checkAutoDetect(comp) {
-			if err = h.applyHelmModuleResources(ctx, comp, owners); err != nil {
-				return errors.Wrap(err, "cannot apply Helm module resources")
-			}
-			continue
-		}
-
-		newComp := comp.DeepCopy()
-		// newComp will be updated and return the revision name instead of the component name
-		revisionName, err := h.createOrUpdateComponent(ctx, newComp)
-		if err != nil {
-			return err
-		}
-		for i := 0; i < len(ac.Spec.Components); i++ {
-			// update the AC using the component revision instead of component name
-			// we have to make AC immutable including the component it's pointing to
-			if ac.Spec.Components[i].ComponentName == newComp.Name {
-				ac.Spec.Components[i].RevisionName = revisionName
-				ac.Spec.Components[i].ComponentName = ""
+		if len(comp.WorkloadAddonResources) != 0 {
+			if _, err := d.Dispatch(ctx, comp.WorkloadAddonResources); err != nil {
+				return errors.WithMessage(err, "cannot dispatch workload addon resources' manifests")
 			}
 		}
-		// isNewRevision indicates app's newly created or spec has changed
-		// skip applying helm resources if no spec change
-		if h.isNewRevision && comp.Spec.Helm != nil {
-			if err = h.applyHelmModuleResources(ctx, comp, owners); err != nil {
-				return errors.Wrap(err, "cannot apply Helm module resources")
-			}
+		if comp.Workload == nil || (len(comp.Workload.GetAPIVersion()) == 0 && len(comp.Workload.GetKind()) == 0) {
+			// TODO implement auto-detect mechanism
+			// this is a auto-detected component, its workload has no GVK so we cannot dispatch its workload and traits
+			// currently only Helm-based component is possible to be auto-detected
+			return fmt.Errorf("helm mode component doesn't specify workload, the traits attached to the helm mode component will fail to work")
```

Comment:
why it's related with trait?

Reply:
it works in cue template. if cue's `output` is empty, traits have no targets to apply.